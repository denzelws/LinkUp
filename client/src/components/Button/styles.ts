import styled, { css, DefaultTheme } from 'styled-components'

import { lighten, darken } from 'polished'
import { ButtonProps } from '.'

export type WrapperProps = Pick<ButtonProps, 'size'>

const wrapperModifiers = {
  small: (theme: DefaultTheme) => css`
    padding: 0 ${theme.spacings.xsmall};
  `,
  fullWidth: () => css`
    width: 100%;
  `
}

export const Wrapper = styled.button<WrapperProps>`
  ${({ theme, size }) => css`
    height: 5.2rem;
    border: none;
    background: #5061fc;
    font-size: ${theme.font.sizes.medium};
    box-shadow: 0 0.4rem 1rem rgba(0, 0, 0, 0.04);
    font-weight: bold;
    color: #fff;
    border-radius: 0.4rem;
    transition: background 0.2s ease-in;
    cursor: pointer;

    &:hover {
      background: ${lighten(0.2, '#5061fc')};
    }

    &:active {
      background: ${darken(0.1, '#5061fc')};
    }

    &[disabled] {
      background: #ccc;
      cursor: default;
    }

    ${!!size && wrapperModifiers[size](theme)}
  `}
`
