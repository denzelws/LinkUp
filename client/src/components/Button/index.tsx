import * as S from './styles'

export type ButtonProps = {
  children: React.ReactNode
  disabled?: boolean
}

const Button = ({ children, disabled }: ButtonProps) => (
  <S.Wrapper disabled={disabled}>{children}</S.Wrapper>
)

export default Button
