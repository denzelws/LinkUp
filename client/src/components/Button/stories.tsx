import { Story, Meta } from '@storybook/react/types-6-0'
import Button, { ButtonProps } from '.'

export default {
  title: 'Button',
  component: Button,
  argTypes: {
    children: {
      type: 'string'
    }
  }
} as Meta

export const Default: Story<ButtonProps> = (args) => (
  <div style={{ maxWidth: '30rem', margin: '0 auto' }}>
    <Button {...args}>Salvar alterações</Button>
  </div>
)

Default.args = {
  children: 'Salvar alterações'
}

export const DisabledButton: Story<ButtonProps> = (args) => (
  <div style={{ maxWidth: '30rem', margin: '0 auto' }}>
    <Button disabled {...args}>
      Salvar alterações
    </Button>
  </div>
)

DisabledButton.args = {
  children: 'Salvar alterações'
}
