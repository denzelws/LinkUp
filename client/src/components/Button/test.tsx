import { screen } from '@testing-library/react'

import Button from '.'
import { renderWithTheme } from '../../utils/tests/helpers'

describe('<Button />', () => {
  it('should render the button', () => {
    const { container } = renderWithTheme(<Button>Test button</Button>)

    expect(
      screen.getByRole('button', { name: /test button/i })
    ).toBeInTheDocument()

    expect(container.firstChild).toMatchSnapshot()
  })

  it('should render disabled button', () => {
    renderWithTheme(<Button disabled>Test button</Button>)

    expect(screen.getByRole('button', { name: /test button/i })).toHaveStyle({
      background: '#ccc',
      cursor: 'default'
    })
  })
})
