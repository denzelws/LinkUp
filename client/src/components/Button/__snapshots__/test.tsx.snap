// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Button /> should render the button 1`] = `
.c0 {
  width: 100%;
  height: 5.2rem;
  border: none;
  background: #5061fc;
  font-size: 1.6rem;
  box-shadow: 0 0.4rem 1rem rgba(0,0,0,0.04);
  font-weight: bold;
  color: #fff;
  border-radius: 0.4rem;
  -webkit-transition: background 0.2s ease-in;
  transition: background 0.2s ease-in;
  cursor: pointer;
}

.c0:hover {
  background: #b4bcfe;
}

.c0:active {
  background: #1e34fb;
}

.c0[disabled] {
  background: #ccc;
  cursor: default;
}

<button
  class="c0"
>
  Test button
</button>
`;
